README.md

```diff
@@ -0,0 +1,453 @@
+  const itemViewMatch = location.pathname.match(itemViewRegex);
+  const itemsMatch = location.pathname.match(itemsRegex);
+
+  if (itemViewMatch) {
+    return { view: 'itemView'; itemId: itemViewMatch[1] };
+  } else if (itemsMatch) {
+    return { view: 'items' };
+  } else {
+    return { view: 'home' };
+  }
+}
+
+export function viewToLocation(view: CurrentView): HistoryLocation {
+  switch (view.view) {
+    case 'itemView':
+      return { pathname: `/items/${view.itemId}`, search: {} };
+    case 'items':
+      return { pathname: '/items', search: {} };
+    case 'home':
+      return { pathname: '/home', search: {} };
+  }
+}
+
+const currentView: ObservableQuery = getCurrentView(locationToView);
+export const doUpdateCurrentView: Command = getDoUpdateCurrentView(viewToLocation);
+```
+
+once you instantiated all the boilerplate needed to instruct Avenger on how to navigate, you can use `currentView` and `doUpdateCurrentView` like they were normal queries and commands (and, in fact, they are..).
+
+```ts
+// ./App.ts
+
+import { declareQueries } from 'avenger/lib/react';
+
+const queries = declareQueries({ currentView });
+
+// usually at the top level of your app there will be a sort of index of your navigation
+class Navigation extends React.PureComponent<Props, State> {
+  render() {
+    return this.props.queries.fold(
+      () => <p>loading</p>,
+      () => null,
+      ({ currentView }) => {
+        switch(currentView.view) {
+          case 'itemView':
+            return <ItemView id={view.itemId} />
+          case 'items':
+            return <Items />
+          case 'home':
+            return <Home />
+        }
+      }
+    )
+  }
+}
+
+export queries(MyComponent)
+
+
+
+// ./Components/ItemView.ts
+
+class ItemView extends React.PureComponent<Props, State> {
+  goToItems: () => doUpdateCurrentView({ view: 'items' }).run()
+
+  render() {
+    return <BackButton onClick={this.goToItems}>
+  }
+}
+
+
+```
+
+## Signatures
+
+> N.B. all the following signatures reference the abstractions in [`fp-ts`](https://github.com/gcanti/fp-ts)
+
+### `query`
+```ts
+declare function query<A = void, L = unknown, P = unknown>(
+  fetch: Fetch<A, L, P>
+): (strategy: Strategy<A, L, P>) => CachedQuery<A, L, P>
+```
+
+### `Fetch`
+
+```ts
+type Fetch<A, L, P> = (input: A) => TaskEither<L, P>;
+```
+
+### `StrategyBuilder`
+
+```ts
+type StrategyBuilder<A, L, P> = (
+  inputSetoid: Setoid<A>,
+  cacheValueSetoid: Setoid<CacheValue<L, P>>
+) => Strategy<A, L, P>;
+```
+
+### `Strategy`
+
+```ts
+export class Strategy<A, L, P> {
+  constructor(
+    readonly inputSetoid: Setoid<A>,
+    readonly filter: Function1<CacheValue<L, P>, boolean>,
+    readonly cacheValueSetoid: Setoid<CacheValue<L, P>>
+  ) {}
+}
+```
+
+### `CachedQuery`
+
+```ts
+interface CachedQuery<A, L, P> {
+  type: 'cached';
+  inputSetoid: Setoid<A>;
+  run: Fetch<A, L, P>;
+  invalidate: Fetch<A, L, P>;
```

Comment:
Should `invalidate` be explained somehow?

Reply:
although exposed, `invalidate` uses are very rare and probably do not deserve to be mentioned here